applications/tari_conduit/src/main.rs

```diff
@@ -0,0 +1,630 @@
+        coinbase_tx: block.miner_tx,
+    });
+}
+
+fn get_seed_hash(data: &[u8]) -> String {
+    let parsed = json::parse(&stringify_request(data)).unwrap();
+    let seed = &parsed["result"]["seed_hash"].to_string();
+    seed.to_owned()
+}
+
+// Temporary till RPC call is in place
+fn add_coinbase(
+    consensus: ConsensusManager,
+    grpc_block: grpc::NewBlockTemplate,
+) -> Result<grpc::NewBlockTemplate, Error>
+{
+    let mut block: NewBlockTemplate = NewBlockTemplate::try_from(grpc_block).unwrap();
+    let fees = block.body.get_total_fee();
+    let (key, r) = get_spending_key()?;
+    let factories = CryptoFactories::default();
+    let builder = CoinbaseBuilder::new(factories);
+    let builder = builder
+        .with_block_height(block.header.height)
+        .with_fees(fees)
+        .with_nonce(r)
+        .with_spend_key(key);
+    let (tx, _unblinded_output) = builder.build(consensus).expect("invalid constructed coinbase");
+    block.body.add_output(tx.body.outputs()[0].clone());
+    block.body.add_kernel(tx.body.kernels()[0].clone());
+    Ok(grpc::NewBlockTemplate::try_from(block).unwrap())
+}
+
+// Temporary till RPC call is in place
+pub fn get_spending_key() -> Result<(PrivateKey, PrivateKey), Error> {
+    let r = PrivateKey::random(&mut OsRng);
+    let key = PrivateKey::random(&mut OsRng);
+    Ok((key, r))
+}
+
+// Add merge mining tag to response
+fn add_merge_mining_tag(data: &[u8], hash: &[u8]) -> Vec<u8> {
+    // Parse the JSON
+    let mut parsed = json::parse(&stringify_request(data)).unwrap();
+
+    // Decode and dserialize the blocktemplate_blob
+    let block_template_blob = &parsed["result"]["blocktemplate_blob"];
+    let s = format!("{}", block_template_blob);
+    let hex = hex::decode(s).unwrap();
+    let mut block = deserialize::<Block>(&hex[..]).unwrap();
+    let tag = Hash(from_slice(hash));
+    let mm_tag = SubField::MergeMining(VarInt(0), tag);
+    block.miner_tx.prefix.extra.0.push(mm_tag);
+
+    // Serialize, encode and replace blocktemplate_blob
+    let serialized = serialize::<Block>(&block);
+    let encode = hex::encode(&serialized);
+    parsed["result"]["blocktemplate_blob"] = encode.into();
+
+    // Recalculate blockhashing blob, then replace it
+    // Formula = append(serialize(blockheader),merkle_root,1+block.tx_hashes.len())
+    // merkle_root =
+    // {
+    //   l = vec![block.miner_tx.hash()]
+    //   l.push(block.tx_hashes)
+    //   tree_hash(l)
+    // }
+    let header = serialize::<BlockHeader>(&block.header);
+    let mut count = serialize::<VarInt>(&VarInt(1 + block.tx_hashes.len() as u64));
+    let mut hashes = block.clone().tx_hashes;
+    hashes.push(block.miner_tx.hash());
+    let mut root = tree_hash(hashes);
+    let mut encode2 = header;
+    encode2.append(&mut root);
+    encode2.append(&mut count);
+    let hex = hex::encode(encode2).into();
+    parsed["result"]["blockhashing_blob"] = hex;
+    parsed.dump().into()
+}
```

Comment:
duplicate

Reply:
Only exists in the conduit?

Reply:
Not as a function, but the logic is already in tari::core. 
But some of the code is def there. 
For example: merged_mining_subfield constructs the mm_tag. But you use the same logic here to create one from scratch again. 
And
You recreate the logic of create_input_blob as well. 


we need to make sure logic like this only happens in a single function and from a single place. That should all be from tari::core

Reply:
True. Will refactor